### PR TITLE
ci: add type checks

### PR DESCRIPTION
**Pull request by JaeAeich**
**IMPORTANT: Please create an issue before filing a pull request! Changes need to be discussed before proceeding. Please read the [contribution guidelines](CONTRIBUTING.md).**

**Details**
This PR add type checking using mypy as stated in issue #77  

**Closing issues**

Fixes issue - #77

### CODE REVIEW

**Comment by JaeAeich** [comment]
@uniqueg Is this good (in right dir)? I still need to narrow it down a little as I have ignored `import errors` in `mypy.ini` which I think needs a little more work.

File: .github/workflows/main.yml [hunk 1]
```diff
@@ -28,6 +28,8 @@ jobs:
         run: pylint pro_wes/ setup.py
       - name: Check code format with Black
         run: black --check setup.py pro_wes/ tests/
+      - name: Type check with mypy
```
Comment (uniqueg) [inline .github/workflows/main.yml]: To be more consistent with the other descriptions, perhaps better to write: "Check types with mypy"

File: mypy.ini [hunk 2]
```diff
@@ -0,0 +1,6 @@
+[mypy]
+ignore_missing_imports = True
```
Comment (uniqueg) [inline mypy.ini:4]: As you already pointed out yourself, I would hesitate to globally disable errors, unless they really affect the entire codebase in almost every module. Otherwise, I think it's better to disable checks one by one. That is also a better reminder to possibly sort these out some time in the future, as it is more likely that one stumbles across ignore statements in the code base, as compared to a config file that nobody looks at much.

And ignoring missing imports globally, in particular, can have quite far reaching consequences, as it will prevent mypy from catching some really serious issues.

File: pro_wes/app.py [hunk 3]
```diff
@@ -32,13 +33,14 @@ def _setup_first_start(app: App) -> None:
         work_dir = Path(current_app.config.foca.custom.post_runs.storage_path.resolve())
         work_dir.mkdir(parents=True, exist_ok=True)
         # set service info
-        service_info = ServiceInfo()
+        service_info: Dict  # pylint: disable=unused-variable
```
Comment (uniqueg) [inline pro_wes/app.py]: I would remove this line (see next comment)

File: pro_wes/app.py [hunk 4]
```diff
@@ -32,13 +33,14 @@ def _setup_first_start(app: App) -> None:
...
         try:
-            service_info = service_info.get_service_info()
+            service_info = ServiceInfo().get_service_info()
```
Comment (uniqueg) [inline pro_wes/app.py]: As it looks like we don't do anything with the response of the call and are only testing if a service info object is available, I think you don't need to assign the response to a variable. And if you remove the assignment, you can also safely remove the variable declaration above. As a side effect, Pylint won't complain about declaring a variable that we don't use.

File: pro_wes/app.py [hunk 5]
```diff
@@ -32,13 +33,14 @@ def _setup_first_start(app: App) -> None:
...
         except NotFound:
-            service_info.set_service_info(
-                data=current_app.config.foca.custom.service_info.dict()
-            )
+            service_info_data: Dict = current_app.config.foca.custom.service_info.dict()
```
Comment (uniqueg) [inline pro_wes/app.py]: I would just call the variable `_data`. It's really just a throwaway variable. Alternatively, you could avoid assigning this entirely and use `current_app.config.foca.custom.service_info.dict()` as is in the `.set_service_info()` call, like so:

```py
            service_info_object.set_service_info(
                data=current_app.config.foca.custom.service_info.dict()
            )
```

One statement less...

File: pro_wes/app.py [hunk 6]
```diff
@@ -32,13 +33,14 @@ def _setup_first_start(app: App) -> None:
...
+            service_info_object = ServiceInfo()
+            service_info_object.set_service_info(data=service_info_data)
+            service_info = service_info_object.get_service_info()  # noqa: F841
```
Comment (uniqueg) [inline pro_wes/app.py]: Again, since we are not making use of the response and just check whether we _now_ have a service info object is available after setting it, you don't need to assign it to a variable. So just call a bare `service_info_object.get_service_info()` without the assignment and remove the Pylint disabling comment.

Also, given that with the suggested changes applied we will not have a `service_info` variable anymore, I think you can rename `service_info_object` to just `service_info` to make the lines a little shorter overall (without losing clarity).

File: pro_wes/app.py [hunk 7]
```diff
@@ -32,13 +33,14 @@ def _setup_first_start(app: App) -> None:
...
+        service_info: Dict  # pylint: disable=unused-variable
         try:
-            service_info = service_info.get_service_info()
+            service_info = ServiceInfo().get_service_info()
         except NotFound:
-            service_info.set_service_info(
-                data=current_app.config.foca.custom.service_info.dict()
-            )
+            service_info_data: Dict = current_app.config.foca.custom.service_info.dict()
+            service_info_object = ServiceInfo()
```
Comment (uniqueg) [inline pro_wes/app.py]: Please add the type hint for `service_info_object`.

File: pro_wes/ga4gh/wes/models.py [hunk 8]
```diff
@@ -38,10 +38,12 @@ class Attachment(BaseModel):
 
     Args:
         filename: Name of the file as indicated in the run request.
+        object: File object.
         path: Path to the file on the app's storage system.
     """
 
     filename: str
+    object: bytes
```
Comment (uniqueg) [inline pro_wes/ga4gh/wes/models.py:46]: Is this really related to type checking? Looks like it might be a separate concern (an incomplete model), so perhaps should go in a different PR? But if it's really mypy complaining about this somewhere, then leave it in here.

File: pro_wes/ga4gh/wes/workflow_runs.py [hunk 9]
```diff
@@ -96,13 +96,22 @@ def run_workflow(
...
         # instantiate WES client
         wes_client: WesClient = WesClient(
             host=document_stored.wes_endpoint.host,
             base_path=document_stored.wes_endpoint.base_path,
             token=kwargs.get("jwt", None),
         )
 
+        if document_stored.task_id is None:
+            raise ValueError("No task ID available")
+
```
Comment (uniqueg) [inline pro_wes/ga4gh/wes/workflow_runs.py]: Hmm, I see that the type for `document_stored` is not indicated. If you add a type hint for that, do we still need these `None` checks for mypy?

And if so, perhaps it's worth adding one or more custom exceptions (or, if applicable, reuse the custom exceptions already defined)? Custom exceptions are defined in `foca.exceptions`, i.e., [here](https://github.com/elixir-cloud-aai/proWES/blob/dev/pro_wes/exceptions.py)). Please note that if you define custom exceptions, you also need to include them in the `exceptions` dictionary in the same module, and add the desired error code and message to be returned to the user as a JSON object.

File: pro_wes/ga4gh/wes/workflow_runs.py [hunk 10]
```diff
@@ -136,25 +145,34 @@ def run_workflow(
...
             raise InternalServerError
-        document_stored: DbDocument = db_connector.upsert_fields_in_root_object(
+        updated_document_stored: DbDocument = db_connector.upsert_fields_in_root_object(
             root="wes_endpoint",
             run_id=response.run_id,
         )
 
+        if updated_document_stored.wes_endpoint is None:
+            raise ValueError("No WES endpoint available.")
+
```
Comment (uniqueg) [inline pro_wes/ga4gh/wes/workflow_runs.py]: Again, a custom error (already available or new) might be useful here.

File: pro_wes/ga4gh/wes/workflow_runs.py (hunk 9 above)
Comment (uniqueg) [reply pro_wes/ga4gh/wes/workflow_runs.py]: Just FYI: Any error that is raised and that is _not_ listed specifically in the `exceptions` dictionary (e.g., `ValueError`) will be automatically routed to `InternalServerError`, with a `500` response code and a generic, uninformative `An unexpected error occurred` message that isn't overly helpful to the user. Hence, custom errors are to be preferred.

File: pro_wes/ga4gh/wes/workflow_runs.py [hunk 11]
```diff
@@ -136,25 +145,34 @@ def run_workflow(
...
+            task_id=updated_document_stored.task_id,
             soft_time_limit=controller_config.timeout_job,
         )
 
-        return {"run_id": document_stored.run_log.run_id}
+        if updated_document_stored.wes_endpoint is None:
+            raise ValueError("No WES endpoint available.")
+
+        if updated_document_stored.wes_endpoint.run_id is None:
+            raise ValueError("WES endpoint does not have run_id.")
```
Comment (uniqueg) [inline pro_wes/ga4gh/wes/workflow_runs.py]: I'm not too happy with all of these checks (here and above) but cannot think of a better solution wihtout diving into the code. Perhaps defining some validator method somewhere to check the content of the `.wes_endpoint` attribute? Then we'd just need to call that validator (which would either raise an error if any of the values were `None` or otherwise return nothing).

The question is whether mypy will scan deep enough to be satisfied with such an approach. If not, what we can always do if _we_ know for sure that, e.g., a given value can't really be `None` (because we have just checked it and mypy failed to process that), is to just add lines like `assert variable is not None`. It's a bit ugly, but it will satisfy mypy and has the advantage that we don't need to write unit tests to check a conditional branch for something that really can't/shouldn't happen.

File: pro_wes/ga4gh/wes/workflow_runs.py [hunk 12]
```diff
@@ -406,7 +424,7 @@ def _create_run_environment(
             # populate document
             document.run_log.run_id = run_id
             document.task_id = uuid()
-            document.work_dir = str(work_dir)
+            document.work_dir = work_dir.as_posix()
```
Comment (uniqueg) [inline pro_wes/ga4gh/wes/workflow_runs.py]: I'm curious: Did this trip up mypy or was suggested by it?

If anybode wanted to run proWES on Windows, this would cause problems.

An example from the [documentation](https://docs.python.org/3/library/pathlib.html#pathlib.PurePath.as_posix)):

```py
p = PureWindowsPath('c:\\windows')
str(p)
# 'c:\\windows'
p.as_posix()
# 'c:/windows'
```

`c:\\windows` and `c:/windows` are _not_ equivalent, obviously.

Of course, trying to run proWES on Windows is probably not a good idea, and I have no clue how Docker handles Windows file paths on Windows (if at all?). But I'm just wondering why `str()` wouldn't be just fine here (and it wouldn't assume anything about the way the file path notation of the underlying OS).

File: pro_wes/ga4gh/wes/workflow_runs.py [hunk 13]
```diff
@@ -437,18 +455,20 @@ def _process_attachments(self, work_dir: Path) -> List[Attachment]:
...
         for file in files:
-            attachments.append(
-                Attachment(
-                    filename=file.filename,
-                    object=file.stream,
-                    path=str(
-                        work_dir
+            if file is not None:
+                if file.filename is None:
+                    raise ValueError("File does not have a filename.")
```
Comment (uniqueg) [inline pro_wes/ga4gh/wes/workflow_runs.py]: I would use assert here to avoid the branching. I don't know exactly under what conditions `file.filename` could be `None`, but if we will ever run into this issue, then an `AssertionError` will be just as informative as the `ValueError` you propose. And then we can still look at the cause of the issue in detail and find a solution for that (e.g., create a random file name or whatever).

File: requirements_dev.txt [hunk 14]
```diff
@@ -5,3 +5,6 @@ mongomock>=4.1.2,<5
 pylint>=2.15.5,<3
 pytest>=7.2.0,<8
 python-semantic-release>=7.32.2,<8
+mypy
```
Comment (uniqueg) [inline requirements_dev.txt]: Would be good to indicate at least a minimum version (the one you tested with). You could, for example, use the comptability matcher (`~=`) on the minor version (e.g., `mypy~=1.8`) to match all `1.x` versions that are at least `1.8.0` (i.e., equivalent to `mypy>=1.8,==1.*`) or you could just pin the minimum version (`mypy>=1.8.0`).

File: setup.py [hunk 15]
```diff
@@ -16,7 +16,7 @@
 
 setup(
     name="pro-wes",
-    version=__version__,  # noqa: F821  # pylint: disable=undefined-variable
+    version=__version__,  # type: ignore # noqa: E501 F821 # pylint: disable=undefined-variable
```
Comment (uniqueg) [inline setup.py:19]: :rofl:

File: pro_wes/tasks/track_run_progress.py [hunk 16]
```diff
@@ -97,7 +97,7 @@ def task__track_run_progress(
     response.pop("request", None)
     document: DbDocument = db_client.upsert_fields_in_root_object(
         root="run_log",
-        **response,
+        **dict(response),
```
Comment (uniqueg) [inline pro_wes/tasks/track_run_progress.py]: Perhaps adding a type hint where `response` is created would prevent the need to cast it to a dictionary? Same for the equivalent change below

File: pro_wes/tasks/track_run_progress.py [hunk 17]
```diff
@@ -106,7 +106,13 @@ def task__track_run_progress(
...
         sleep(controller_config.polling_wait)
         try:
-            response = wes_client.get_run_status(
+            if document.wes_endpoint is None:
+                raise ValueError("WES run ID not available.")
+
+            if document.wes_endpoint.run_id is None:
+                raise ValueError("WES run ID not available.")
+
+            wes_client.get_run_status(
```
Comment (uniqueg) [inline pro_wes/tasks/track_run_progress.py]: Check other comments about custom errors and possibly using another approach to validate `.wes_endpoint`

File: pro_wes/tasks/track_run_progress.py [hunk 18]
```diff
@@ -133,6 +139,9 @@ def task__track_run_progress(
     try:
         # workaround for cwl-WES; add .dict() when cwl-WES response conforms
         # to model
+        if response.run_id is None:
+            raise ValueError("WES run ID not available.")
+
```
Comment (uniqueg) [inline pro_wes/tasks/track_run_progress.py]: Sure this should go inside this `try` `except` block?

**Review by uniqueg — CHANGES_REQUESTED** [review]
Thanks @JaeAeich. Quite a number of comments, but I guess they are mostly small things (except perhaps the `.wes_endpoint` validation, which, however, is not necessary).

**Comment by JaeAeich** [comment]
Hey @uniqueg, unfortunately `mypy` doesn't check that deep, this is in regard to your suggestion to create a validator func. So the other solution would be something like this:
```
        assert document_stored.wes_endpoint is not None, "No WES endpoint available."
        assert (
            document_stored.wes_endpoint.base_path is not None
        ), "WES endpoint does not have base_path."

        if document_stored.task_id is None:
            raise WesEndpointProblem
```
Where `WesEndpointProblem` is
```
....
class WesEndpointProblem(NotFound):
    """No/few reuirements provided for WES endpoint."""


exceptions={
   ....
    WesEndpointProblem: {
        "message": "No/few reuirements provided for WES endpoint.",
        "code": "500",
    }
}
```
yeah that mean I gotta do it in two place ie, for `document`, `updated_document` and in `track_run_progress`. Please do tell me if there is better way to do it.
PS: Not yet ready for reviews.

File: .github/workflows/main.yml (hunk 1 above)
Comment (JaeAeich) [reply .github/workflows/main.yml]: done

File: mypy.ini (hunk 2 above)
Comment (JaeAeich) [reply mypy.ini:4]: will slowly strip each rule, otherwise CI breaks and I can't seem to figure out what broke it :slightly_smiling_face:

File: pro_wes/app.py (hunk 3 above)
Comment (JaeAeich) [reply pro_wes/app.py]: BTW, Removes all this, because `service_info` doesn't seem to be doing anything in terms of storing the value, its just for setting `service_info`  (as you mentioned below), in that case we can just remove the definition and just have something like
```
        # set service info
        try:
            ServiceInfo().get_service_info()
        except NotFound:
            ServiceInfo().set_service_info(
                data=current_app.config.foca.custom.service_info.dict()
            )
```
do correct me if I misunderstood :slightly_smiling_face:

File: requirements_dev.txt (hunk 14 above)
Comment (JaeAeich) [reply requirements_dev.txt]: done

File: pro_wes/ga4gh/wes/models.py (hunk 8 above)
Comment (JaeAeich) [reply pro_wes/ga4gh/wes/models.py:46]: yeah `mypy` is complaining, hence will need to be in this PR ig.

File: pro_wes/ga4gh/wes/workflow_runs.py (hunk 12 above)
Comment (JaeAeich) [reply pro_wes/ga4gh/wes/workflow_runs.py]: yeah `mypy` did complain, I have ignored the error,
```
error: Incompatible types in assignment (expression has type "str | None", variable has type "Path | None")
```
Idk for some reason, even checks aren't working.

File: pro_wes/ga4gh/wes/workflow_runs.py [hunk 19]
```diff
@@ -424,7 +427,7 @@ def _create_run_environment(
             # populate document
             document.run_log.run_id = run_id
             document.task_id = uuid()
-            document.work_dir = work_dir.as_posix()
+            document.work_dir = str(work_dir)  # type: ignore
```
Comment (uniqueg) [inline pro_wes/ga4gh/wes/workflow_runs.py]: Sorry, but I'm a bit confused. What's the expected type for `document.work_dir`, and why don't you assign it the type it wants?

**Review by uniqueg — CHANGES_REQUESTED** [review]
Good stuff! Just one question I still have (but maybe I'm missing something)

**Comment by uniqueg** [comment]
> Hey @uniqueg, unfortunately `mypy` doesn't check that deep, this is in regard to your suggestion to create a validator func. So the other solution would be something like this:
> 
> ```
>         assert document_stored.wes_endpoint is not None, "No WES endpoint available."
>         assert (
>             document_stored.wes_endpoint.base_path is not None
>         ), "WES endpoint does not have base_path."
> 
>         if document_stored.task_id is None:
>             raise WesEndpointProblem
> ```
> 
> Where `WesEndpointProblem` is
> 
> ```
> ....
> class WesEndpointProblem(NotFound):
>     """No/few reuirements provided for WES endpoint."""
> 
> 
> exceptions={
>    ....
>     WesEndpointProblem: {
>         "message": "No/few reuirements provided for WES endpoint.",
>         "code": "500",
>     }
> }
> ```
> 
> yeah that mean I gotta do it in two place ie, for `document`, `updated_document` and in `track_run_progress`. Please do tell me if there is better way to do it. PS: Not yet ready for reviews.

Looks fine to me, not too messy :)

**Comment by uniqueg** [comment]
Sorry, only saw now that you were not ready for review. Gave you one anyway :laughing:

But it seems to me that you addressed at least most of my previous issues. Please make sure to resolve the corresponding conversations so that we can better keep track of what's still left to do.

**Comment by JaeAeich** [comment]
```
ERROR: pip's dependency resolver does not currently take into account all the packages that are installed. This behaviour is the source of the following dependency conflicts.
foca 0.12.1 requires pymongo==3.10.1, but you have pymongo 3.13.0 which is incompatible.
```

I am disabling all error for missing imports, when `foca` update `pymongo` to `3.11` we can remove it. This is because 
`pymongo-stubs is compatible with Python >=3.6 and [PyMongo](https://pypi.org/project/pymongo/) >=3.11,<4.0.`

File: mypy.ini (hunk 2 above)
Comment (JaeAeich) [reply mypy.ini:4]: @uniqueg I am left with these errors for now, since most are dependent on `foca` how do I go by slowing them supressing `attr-defined` globally doesn't seem to be right and I can't define this in `mypy.ini` all i can do is `# type: ignore`, should I do crazy with `# type: ignore` for these errors?

```
pro_wes/ga4gh/wes/service_info.py:33: error: "Config" has no attribute "foca"  [attr-defined]
pro_wes/app.py:32: error: "Config" has no attribute "foca"  [attr-defined]
pro_wes/app.py:39: error: "Config" has no attribute "foca"  [attr-defined]
pro_wes/ga4gh/wes/models.py:161: error: "str" has no attribute "name"  [attr-defined]
pro_wes/tasks/track_run_progress.py:60: error: "Config" has no attribute "foca"  [attr-defined]
pro_wes/tasks/track_run_progress.py:98: error: "RunLog" has no attribute "pop"  [attr-defined]
pro_wes/tasks/track_run_progress.py:108: error: "dict[Any, Any]" has no attribute "polling_wait"  [attr-defined]
pro_wes/tasks/track_run_progress.py:120: error: "dict[Any, Any]" has no attribute "polling_attempts"  [attr-defined]
pro_wes/tasks/track_run_progress.py:127: error: "dict[Any, Any]" has no attribute "polling_attempts"  [attr-defined]
pro_wes/tasks/track_run_progress.py:151: error: "RunLog" has no attribute "pop"  [attr-defined]
pro_wes/ga4gh/wes/workflow_runs.py:59: error: "Config" has no attribute "foca"  [attr-defined]
```

File: pro_wes/tasks/track_run_progress.py (hunk 16 above)
Comment (JaeAeich) [reply pro_wes/tasks/track_run_progress.py]: Since response is of type `RunLog`, I am typecasting it, if I declare it to be `Dict` problems arise down the code. `response: RunLog = wes_client.get_run(run_id=remote_run_id)`

File: setup.py (hunk 15 above)
Comment (JaeAeich) [reply setup.py:19]: I have no idea what to do here :slightly_smiling_face:

Comment (uniqueg) [reply setup.py:19]: It's fine, just funny :)

Actually, I think we don't run mypy on `setup.py` in other repos. But you can leave it. Soon we will migrate to `pyproject.toml` anyway and then that will not be an issue anymore.

File: mypy.ini (hunk 2 above)
Comment (uniqueg) [reply mypy.ini:4]: Yes, I'm afraid so - it's what we did in the other repos as well for now.

We will fix this with the upcoming FOCA release, but it will take a couple of weeks. Then we can revisit.

**Comment by JaeAeich** [comment]
@uniqueg Please review the PR. I have made the requested changes :+1:

**Review by uniqueg — APPROVED** [review]
Great - thanks a lot @JaeAeich :)
